### PR TITLE
Provide Connection Strength Indicator for Teacher and Students 

### PR DESCRIPTION
**Pull request by farmanahmed888**
A connection strength indicator would help users proactively address connectivity issues before they impact class quality.
a status bar [like youtube] in the Android app or web app whenever connection to internet is disrupted.

## New Features
- Added real-time network connectivity monitoring that tracks your connection status during calls
- On-screen connectivity banner displays when you're offline or experiencing a weak connection
- Automatic notifications inform you when your connection is restored
- System periodically performs health checks to detect and alert degraded network conditions

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
Below is your complete, single‑response code review following all rules.

---

## High‑Level Summary

The feature is well‑designed across Android, backend, and web. The backend route and tests are clean. However, the Android implementation has **two severe correctness issues** (incorrect Retrofit handling and swallowing real error information) and the web hook has **one logic bug** where it ignores the fetch response entirely.

Several places also violate global error‑handling and logging standards. No security issues detected.

---

## Issues

### CRITICAL — Android Retrofit call not checked for failure  
**File:** ConnectivityRepository.kt  
**Function:** `ping()`  
**Problem:** `service.healthPing()` returns a `Response<Unit>`, but the code assumes success and never checks `isSuccessful`. A 500 or 400 response results in a false “ONLINE/DEGRADED” state and hides backend outages.  
**Fix:** After calling the Retrofit method, explicitly check `response.isSuccessful`; if false, treat it as offline.

Deduction: -8

---

### HIGH — Android exception handler drops context and violates error-logging standards  
**File:** ConnectivityRepository.kt  
**Function:** `ping()`  
**Problem:** Catch block silently maps all exceptions (including cancellation) to OFFLINE without logging error context. Violates team error‑logging rules.  
**Fix:** Log structured ERROR-level message with function name and a summary of input (e.g., elapsed time, state). Then apply fallback.

Deduction: -5

---

### HIGH — Web hook ignores HTTP status, treating all responses as success  
**File:** useConnectivity.js  
**Function:** `ping`  
**Problem:** Fetch result is ignored — even if `res.ok === false`, the hook marks the connection ONLINE/DEGRADED falsely.  
**Fix:** Read the response returned by `fetch` and treat non‑OK responses as offline.

Deduction: -5

---

### MEDIUM — Android: Missing timeout configuration on Retrofit interface  
**File:** ConnectivityRepository.kt / SeedsService  
**Function:** `healthPing()`  
**Problem:** Global rule requires explicit API call timeouts. Current implementation relies on `withTimeout` at the caller, but Retrofit call itself has no explicit timeout configuration.  
**Fix:** Define a dedicated OkHttp client with explicit read/write/connect timeouts for this API.

Deduction: -3

---

### MEDIUM — No structured logging in Android and Web  
**Files:**  
• ConnectivityRepository.kt (`ping`, `applyStatus`)  
• CallFragment.kt (`updateConnectivityBanner`)  
**Problem:** No structured logs around state transitions. UI behaviour dependent on these transitions will be hard to debug.  
**Fix:** Replace `Log.*` usage with structured logs from the team-standard logger.

Deduction: -3

---

### LOW — Hardcoded color values in Android UI  
**File:** CallFragment.kt  
**Function:** `updateConnectivityBanner`  
**Problem:** Magic hex colors violate config centralisation rule.  
**Fix:** Move colors to theme resources.

Deduction: -1

---

### LOW — Missing test coverage for Android repository  
**File:** ConnectivityRepository.kt  
**Problem:** New service‑level logic added without unit tests.  
**Fix:** Add tests verifying degraded/online/offline transitions, debounce behaviour, and network callback behaviour.

Deduction: -1

---

## Scoring

Total deductions = 8 + 5 + 5 + 3 + 3 + 1 + 1 = **26**

Score = 100 – 26 = **74**

---

## Optional References  
None required — no non‑obvious patterns beyond standard Android/Retrofit behavior.

**Review by a4i-architect — COMMENTED** [review]
### High‑Level Summary  
Feature is generally well‑designed (bidirectional connectivity monitoring across Android/web, server ping route, unit tests).  
However, several **CRITICAL / HIGH** issues exist around **incorrect API handling, missing error logging, misuse of Retrofit return types, leaking coroutines, and inconsistent status logic**.  
Multiple coding‑standards violations appear (error handling, logging, config centralization, typing).

Below are all issues, sorted by severity, with file + function + actionable fix.

---

## CRITICAL ISSUES (‑8 pts each)

### 1. **Android: healthPing() return type misused as if it throws on non‑200**  
**File:** `ConnectivityRepository.kt`, function `ping()`  
**Issue:** Retrofit's `suspend fun healthPing(): Response<Unit>` *never throws* on HTTP errors. A non‑200 response will incorrectly be treated as success → leading to false "ONLINE/DEGRADED" states.  
**Fix:**  
Change API to return `Unit` directly using `suspend fun healthPing(): Unit` and enable Retrofit `KotlinxSerializationCallAdapter` OR explicitly check `response.isSuccessful` and treat failures as offline.

---

### 2. **Android: No logging of errors violates global ERROR HANDLING rule**  
**File:** `ConnectivityRepository.kt`, function `ping()` catch block  
**Issue:** Exceptions are swallowed with **zero logging** — direct violation of Error Handling rule #3.  
**Fix:**  
Before `applyStatus(OFFLINE)`, log using structured logging (e.g., Timber) with function name + summary:  
`logger.error("ConnectivityRepository.ping failure", mapOf("error" to e::class.simpleName))`.

---

### 3. **Web: fetch() result ignored — non‑200 treated as success**  
**File:** `useConnectivity.js`, function `ping()`  
**Issue:** `fetch()` resolving with `{ ok: false }` still becomes ONLINE/DEGRADED; same problem as Android. Could mask backend outages.  
**Fix:**  
Check `response.ok`; if false → `applyStatus("offline")`.

---

### 4. **Android: connectivityManager.registerNetworkCallback() never unregistered → memory leak**  
**File:** `ConnectivityRepository.kt`, class init block  
**Issue:** Repository is @Singleton → callback stays registered forever even after app background/kill cycles.  
**Fix:**  
Expose `shutdown()` and call `connectivityManager.unregisterNetworkCallback(networkCallback)` in `onCleared()` or Application shutdown event.

---

## HIGH ISSUES (‑5 pts each)

### 5. **Android: Missing HTTP timeout config violates API rule #1**  
**File:** `ConnectivityRepository.kt`, function `ping()`  
**Issue:** Using Retrofit without client‑side timeouts specified. Only coroutine timeout wraps the call; underlying OkHttp may still slow‑drain connections.  
**Fix:**  
Ensure OkHttp client of `SeedsService` explicitly sets connect/read/write timeouts ≤ `ABORT_TIMEOUT_MS`.

---

### 6. **Android: Magic numbers violate config centralization**  
**File:** `ConnectivityRepository.kt`, companion object constants  
**Issue:** Global rule requires config in centralized config/env.  
**Fix:**  
Move PING_INTERVAL_MS, PING_THRESHOLD_MS, ABORT_TIMEOUT_MS, DEBOUNCE_MS into shared `Config.kt`.

---

### 7. **CallFragment logs using Log.d, violating structured logging requirement**  
**File:** `CallFragment.kt`, functions `setupObservers`, `onStart`, `onStop`  
**Fix:** Replace with structured logging abstraction (e.g., Timber or custom logger).

---

### 8. **Backend /health/ping route not covered by API rate limiter**  
**File:** `backend-server/src/index.js`  
**Issue:** healthRouter is mounted before rate limiter. DDoSable by unbounded pings from clients.  
**Fix:**  
Move `app.use(limiter)` above health router OR add lightweight limiter specifically on `/health/ping`.

---

## MEDIUM ISSUES (‑3 pts each)

### 9. **Android: ping coroutine launched from NetworkCallback is unscoped → potential overlap**  
**File:** `ConnectivityRepository.kt`, callback onAvailable  
**Issue:** Multiple concurrent `ping()` executions possible if network flaps quickly.  
**Fix:**  
Use a mutex or guard flag to ensure only one ping runs at a time.

---

### 10. **Android: No error logging in startSessionMonitoring loop**  
**File:** `ConnectivityRepository.kt`, `startSessionMonitoring()`  
**Issue:** Exceptions inside `ping()` quietly drop future pings if coroutine crashes (depending on SupervisorJob behavior).  
**Fix:**  
Wrap body in try/catch with structured error logging.

---

### 11. **Web: setInterval cleanup doesn’t null intervalRef — double intervals possible**  
**File:** `useConnectivity.js`, effect watching isSessionActive  
**Fix:**  
Set `intervalRef.current = null` after clearing interval.

---

### 12. **Other platform mismatch in status strings ("ONLINE" vs "online")**  
**File:** `useConnectivity.js` and Android enum ConnectivityStatus  
**Issue:** Android uses uppercase enum; web uses lowercase. UI behavior diverges if server someday harmonizes responses.  
**Fix:**  
Standardize names ("ONLINE", "DEGRADED", "OFFLINE").

---

## LOW ISSUES (‑1 pt each)

### 13. **Android: No unit tests for ConnectivityRepository**  
**File:** `ConnectivityRepository.kt`  
**Fix:** Add tests covering degraded/timeout/offline boundaries.

---

### 14. **Web: applyStatus uses nested setState calls but doesn’t cancel prior ping‑induced updates**  
**File:** `useConnectivity.js`, applyStatus()  
**Fix:**  
Store last intended state to avoid race between ping and offline events.

---

### 15. **Backend: ping route missing INFO logging for observability**  
**File:** `healthRouter.js`  
**Fix:**  
Add structured info log: endpoint invoked, no sensitive data.

---

## Final Score Calculation  
CRITICAL (4) → ‑32  
HIGH (4) → ‑20  
MEDIUM (4) → ‑12  
LOW (3) → ‑3  
**Total deduction = 67  
Score = 100 − 67 = 33**

---

If you'd like, I can generate a corrected patch for the highest‑severity issues.

**Review by a4i-architect — COMMENTED** [review]
### High‑Level Summary
Overall implementation is solid, but several **error‑handling, logging, strict typing, and architectural standard violations** exist. A few Android‑side issues may cause **UI memory leaks** or **uncancelled callbacks**, and the backend has a **missing timeout configuration** for external calls. Webapp code mostly meets standards but violates **structured logging requirements** and **centralised config rules**.

Below are all issues, prioritised and fully actionable.

---

## Issues

### CRITICAL (‑8)

#### 1. Missing graceful error logging & context in `ConnectivityRepository.ping()`  
**File:** ConnectivityRepository.kt, function `ping()`  
**Violation:** Error handling rule 3; sensitive logging.  
**Issue:** Exception is swallowed without structured logging. Debugging field failures becomes impossible.  
**Fix:** Add structured log call (non‑PII) inside `catch` before calling `applyStatus`.

#### 2. Missing cancellation of `connectivityManager.registerNetworkCallback()`  
**File:** ConnectivityRepository.kt, class init block  
**Issue:** Callback is never unregistered → leaking the Fragment/Application context.  
**Fix:** Add `fun cleanup()` to call `connectivityManager.unregisterNetworkCallback(networkCallback)` and call it from an appropriate lifecycle (e.g., repository clearing or App exit hook).

---

### HIGH (‑5)

#### 3. No timeout set on `fetch()` in `useConnectivity`  
**File:** useConnectivity.js, function `ping`  
**Violation:** API rule 1.  
**Issue:** You only abort via custom `setTimeout`, but `fetch` may still hang on DNS/connectivity edge cases.  
**Fix:** Use `{ signal: controller.signal }` AND ensure caller has global fetch timeout (AbortController is fine but must wrap fetch in `Promise.race` for robustness).

#### 4. Missing error propagation for `healthPing()`  
**File:** ConnectivityRepository.kt, `ping()`  
**Violation:** Error handling rule 1.  
**Issue:** `service.healthPing()` returns a `Response<Unit>` that may be non‑successful yet not thrown; offline status will be misreported.  
**Fix:** Validate response with `if (!resp.isSuccessful) throw IOException(...)`.

#### 5. Incorrect use of `System.currentTimeMillis()` for latency measurement  
**File:** ConnectivityRepository.kt, `ping()`  
**Issue:** Wall clock changes can corrupt degradation logic.  
**Fix:** Use `SystemClock.elapsedRealtime()` instead.

---

### MEDIUM (‑3)

#### 6. Missing structured logging across new Android/Web code  
**Files:** ConnectivityRepository.kt (multiple), CallFragment.kt (Log.d), useConnectivity.js (none)  
**Violation:** Logging rule 1.  
**Fix:** Replace Log.d with structured logger wrapper; add non‑PII context (function name, event).

#### 7. Magic values violate configuration centralisation  
**Files:** ConnectivityRepository.kt (constants), useConnectivity.js (same constants duplicated)  
**Violation:** Architecture rule 1.  
**Fix:** Move intervals, thresholds to a shared config module.

#### 8. Missing `@Throws` annotation or explicit documentation for `healthPing()` failure path  
**File:** SeedsService.healthPing()  
**Issue:** Consumers may assume silent success.  
**Fix:** Document non‑200 status handling or wrap Retrofit call with `suspend fun healthPingOrThrow()`.

---

### LOW (‑1)

#### 9. CallFragment.observe logs without correct severity  
**File:** CallFragment.kt, function `setupObservers`  
**Issue:** Using debug logs for degraded/offline events that should be WARN.  
**Fix:** Replace with structured logging at appropriate level.

#### 10. Hardcoded hex colors in `updateConnectivityBanner`  
**File:** CallFragment.kt  
**Issue:** Violates theme consistency and config centralisation.  
**Fix:** Move colors to XML resources.

#### 11. Test coverage missing for Android repository  
**File:** ConnectivityRepository.kt  
**Violation:** Testing rule 1.  
**Fix:** Add unit test validating degraded/offline/online transitions.

---

## Score Calculation

CRITICAL: 2 × ‑8 = ‑16  
HIGH: 3 × ‑5 = ‑15  
MEDIUM: 3 × ‑3 = ‑9  
LOW: 3 × ‑1 = ‑3  

Total deduction = **‑43**  
Final Score = **100 – 43 = 57**

---

If you'd like, I can produce the fixed patches for each issue.

**Comment by farmanahmed888** [comment]
> ## 🟡 High: Flaky backend test timing assertion
> **File:** `backend-server/tests/routes/health.test.js:12-15`
> 
> ```js
> it('responds quickly (under 200ms)', async () => {
>     const start = Date.now();
>     await request(app).get('/health/ping');
>     expect(Date.now() - start).toBeLessThan(50);
> });
> ```
> 
> The test description says "under 200ms" but the assertion checks `< 50ms`. A supertest round-trip involves HTTP parsing overhead that can easily exceed 50ms under CI load. This will flake. Either:
> 
> * Raise the threshold to something realistic (e.g., 500ms), or
> * Remove the timing test entirely — the endpoint is a trivial `res.status(200).end()`, so there's no meaningful performance to assert

done

File: backend-server/tests/routes/health.test.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  it('returns 200 with empty body', async () => {
+    const res = await request(app).get('/health/ping');
+    expect(res.status).toBe(200);
+    expect(res.text).toBe("");
+  });
+
+  it('responds quickly (under 200ms)', async () => {
+    const start = Date.now();
+    await request(app).get('/health/ping');
+    expect(Date.now() - start).toBeLessThan(50);
```
Comment (theainerd) [inline backend-server/tests/routes/health.test.js]: 🟡 **High: flaky timing assertion.** The test description says "under 200ms" but the assertion checks `< 50ms`. A supertest round-trip involves HTTP parsing overhead that can easily exceed 50ms under CI load — this will flake.

Either raise the threshold significantly (e.g., 500ms), or remove this timing test entirely — the endpoint is a trivial `res.status(200).end()`, so there's no meaningful performance to assert.

**Review by theainerd — COMMENTED** [review]
One high-priority item: flaky test assertion.

File: backend-server/tests/routes/health.test.js (hunk 1 above)
Comment (farmanahmed888) [reply backend-server/tests/routes/health.test.js]: done

File: backend-server/src/index.js [hunk 2]
```diff
@@ -26,6 +26,9 @@ const TEACHER_ROLE = "teacher";
 // Initialize Swagger
 setupSwagger(app);
 
+const healthRouter = require("./routes/healthRouter");
+app.use("/health", healthRouter);
```
Comment (theainerd) [inline backend-server/src/index.js]: Because this route is registered before `app.use(cors())`, requests to `/health/ping` never pass through the CORS middleware. In the web application cross-origin setup where `REACT_APP_API_BASE_URL` points at the API host, the browser will reject the ping response and `useConnectivity` will mark active sessions offline even when the server returns 200. Mount this after CORS or apply CORS before the health router.

File: teacher-webapp/src/hooks/useConnectivity.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      });
+    }, DEBOUNCE_MS);
+  }, []);
+
+  const ping = useCallback(async () => {
+    const controller = new AbortController();
+    const hardAbort = setTimeout(() => controller.abort(), ABORT_TIMEOUT_MS);
+    const start = Date.now();
+    try {
+      await fetch(API_ENDPOINTS.HEALTH_PING, { signal: controller.signal });
```
Comment (theainerd) [inline teacher-webapp/src/hooks/useConnectivity.js:30]: When `/health/ping` returns an HTTP error such as 404, 502, or 503, `fetch` still resolves normally, so this code measures the response time and reports `online` or `degraded` even though the health check failed. Check the returned response `ok`/status before applying an online status so proxy or backend error responses do not hide outages.

File: Teacher-App/app/src/main/java/com/example/seeds/connectivity/ConnectivityRepository.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            delay(DEBOUNCE_MS)
+            _status.value = next
+        }
+    }
+
+    private suspend fun ping() {
+        val start = System.currentTimeMillis()
+        try {
+            withTimeout(ABORT_TIMEOUT_MS) {
+                service.healthPing()
```
Comment (theainerd) [inline Teacher-App/app/src/main/java/com/example/seeds/connectivity/ConnectivityRepository.kt:90]: Because `healthPing()` returns `Response<Unit>`, Retrofit does not throw for non-2xx responses; a 503, 502, or 404 health response will reach the elapsed-time logic and be reported as `ONLINE` or `DEGRADED`. Inspect `response.isSuccessful` inside the timeout and treat unsuccessful responses as offline so Android does not hide backend or proxy failures.

File: backend-server/src/index.js (hunk 2 above)
Comment (farmanahmed888) [reply backend-server/src/index.js]: done

File: teacher-webapp/src/hooks/useConnectivity.js (hunk 3 above)
Comment (farmanahmed888) [reply teacher-webapp/src/hooks/useConnectivity.js:30]:
```js
router.get("/ping", (req, res) => {
  res.status(200).end();
});
```
this is code for `/health/ping`  
justify how will we get HTTP error such as 404, 502, or 503

File: Teacher-App/app/src/main/java/com/example/seeds/connectivity/ConnectivityRepository.kt (hunk 4 above)
Comment (farmanahmed888) [reply Teacher-App/app/src/main/java/com/example/seeds/connectivity/ConnectivityRepository.kt:90]:
```js
router.get("/ping", (req, res) => {
  res.status(200).end();
});
```
this is code for `/health/ping`  
justify how will we get HTTP error such as 404, 502, or 503

**Review by a4i-architect — COMMENTED** [review]
Below is the full review following all rules.

---

## High‑Level Summary

The feature is sound, but several violations of global standards appear — especially in **error logging**, **fallback behavior**, and **network response validation**. The Android `ConnectivityRepository` is the main area requiring corrections. A few LOW‑severity issues appear in UI code. Backend and webapp additions are clean.

---

## Issues

### 1. **Missing required structured error logging**  
**File:** `ConnectivityRepository.kt`  
**Function:** `ping()`  
**Severity:** MEDIUM (error‑handling standard violation)  

The `catch` block swallows all exceptions without logging any context, violating the global ERROR HANDLING rule requiring contextual logging before propagation or fallback.

**Fix:** Add structured logging (not print/Log.d) inside the catch branch with function name and summary of operation.

---

### 2. **Incorrect handling of Retrofit Response**  
**File:** `ConnectivityRepository.kt`  
**Function:** `ping()`  
**Severity:** HIGH (logic bug)

`service.healthPing()` returns `Response<Unit>`, but the code treats successful *transport* as equivalent to successful *HTTP*. Non‑200 statuses will incorrectly be marked ONLINE/DEGRADED.

**Fix:** After the call, check `response.isSuccessful`; if false, treat as offline and call `applyStatus(OFFLINE)`.

---

### 3. **Missing logging of offline transitions**  
**File:** `ConnectivityRepository.kt`  
**Function:** `networkCallback.onLost()`  
**Severity:** MEDIUM  

The code updates connectivity state but no structured log is written, violating the ERROR HANDLING standard (recoverable but must log context).

**Fix:** Add INFO/WARNING level structured log with context (function + network type summary) without exposing sensitive data.

---

### 4. **Magic values for UI colors**  
**File:** `CallFragment.kt`  
**Function:** `updateConnectivityBanner()`  
**Severity:** LOW (configuration standard)

Raw hex color values violate the rule that configuration values must be centralized.

**Fix:** Move color constants to resources (colors.xml) and reference via `ContextCompat.getColor()`.

---

### 5. **Magic numeric constants in ConnectivityRepository**  
**File:** `ConnectivityRepository.kt`  
**Location:** `companion object` constants  
**Severity:** LOW  

Although grouped, these values (intervals, thresholds) are part of runtime configuration and should be sourced from environment/config files according to ARCHITECTURE requirements.

**Fix:** Move PING_INTERVAL_MS, PING_THRESHOLD_MS, ABORT_TIMEOUT_MS, DEBOUNCE_MS into a central config provider.

---

### 6. **Android networkCallback lacks unregister call** (pre‑existing but worsened by new usage)  
**File:** `ConnectivityRepository.kt`  
**Function:** `init { ... }`  
**Severity:** MEDIUM  

This PR increases reliance on the callback but never unregisters it, creating potential memory leaks and duplicated callbacks when the repo is recreated (possible depending on Hilt scope).

**Fix:** Provide an explicit `close()` or Hilt‑controlled lifecycle to call `connectivityManager.unregisterNetworkCallback(networkCallback)`.

---

## Score Calculation

- HIGH: 1 issue → -3  
- MEDIUM: 3 issues → -6  
- LOW: 2 issues → -2  

Total deductions = 11  
**Score = 100 - 11 = 89**

---

## References (non‑obvious behavior)

- Retrofit Response success rules: https://square.github.io/retrofit/2.x/retrofit/retrofit2/Response.html  
- Android ConnectivityManager callback lifecycle: https://developer.android.com/reference/android/net/ConnectivityManager#unregisterNetworkCallback(android.net.ConnectivityManager.NetworkCallback)

---

Let me know if you want the corrected patches generated automatically.

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt [hunk 5]
```diff
@@ -271,6 +271,26 @@ class CallFragment : BaseFragment() {
...
+    private fun updateConnectivityBanner(status: ConnectivityStatus) {
+        val container = binding.connectivityBannerContainer
+        val text = binding.connectivityBannerText
+        when (status) {
+            ConnectivityStatus.OFFLINE -> {
+                container.setBackgroundColor(Color.parseColor("#B00020"))
+                text.text = getString(R.string.connectivity_offline)
+                container.visibility = View.VISIBLE
+            }
+            ConnectivityStatus.DEGRADED -> {
```
Comment (theainerd) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt:283]: Nit: hardcoded colors (`#B00020`, `#E65100`) should go in `colors.xml` for theming/dark mode support. Fine as a follow-up.

File: teacher-webapp/src/hooks/useConnectivity.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    };
+  }, [applyStatus, ping]);
+
+  useEffect(() => {
+    if (!isSessionActive) {
+      clearInterval(intervalRef.current);
+      return;
+    }
+    intervalRef.current = setInterval(ping, PING_INTERVAL_MS);
+    return () => clearInterval(intervalRef.current);
```
Comment (theainerd) [inline teacher-webapp/src/hooks/useConnectivity.js:57]: Minor: this interval runs even when the browser tab is hidden, which wastes battery on mobile. Consider pausing when `document.hidden === true`. Non-blocking — follow-up is fine.

**Review by theainerd — APPROVED** [review]
Looks good overall. Clean cross-platform feature with solid test coverage. Minor suggestions as inline comments — none are blockers.

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt (hunk 5 above)
Comment (farmanahmed888) [reply Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt:283]: no dark mode as of yet

File: teacher-webapp/src/hooks/useConnectivity.js (hunk 6 above)
Comment (farmanahmed888) [reply teacher-webapp/src/hooks/useConnectivity.js:57]: can you provide metrics on how much battery does one ping to server every x seconds will it consume?

**Review by N1kunj1998 — APPROVED** [review]
LGTM
